lucene/analysis/smartcn/src/java/org/apache/lucene/analysis/cn/smart/hhmm/BigramDictionary.java

```diff
@@ -0,0 +1,265 @@
+          throw new RuntimeException(ioe);
+        }
+      } catch (ClassNotFoundException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    return singleInstance;
+  }
+
+  private boolean loadFromObj(Path serialObj) {
+    try {
+      loadFromInputStream(Files.newInputStream(serialObj));
+      return true;
+    } catch (Exception e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @SuppressForbidden(reason = "TODO: fix code to serialize its own dictionary vs. a binary blob in the codebase")
+  private void loadFromInputStream(InputStream serialObjectInputStream)
+      throws IOException, ClassNotFoundException {
+    try (ObjectInputStream input = new ObjectInputStream(serialObjectInputStream)) {
+      bigramHashTable = (long[]) input.readObject();
+      frequencyTable = (int[]) input.readObject();
+      // log.info("load bigram dict from serialization.");
+    }
+  }
+
+  @SuppressForbidden(reason = "TODO: fix code to serialize its own dictionary vs. a binary blob in the codebase")
+  private void saveToObj(Path serialObj) throws IOException {
+    try (ObjectOutputStream output = new ObjectOutputStream(Files.newOutputStream(
+        serialObj))) {
+      output.writeObject(bigramHashTable);
+      output.writeObject(frequencyTable);
+      // log.info("serialize bigram dict.");
+    }
+  }
+
+  private void load() throws IOException, ClassNotFoundException {
+    InputStream input = this.getClass().getResourceAsStream("bigramdict.mem");
+    loadFromInputStream(input);
+  }
+
+  private void load(String dictRoot) throws IOException {
+    String bigramDictPath = dictRoot + "/bigramdict.dct";
+
+    Path serialObj = Paths.get(dictRoot + "/bigramdict.mem");
+
+    if (Files.exists(serialObj) && loadFromObj(serialObj)) {
+
+    } else {
+      try {
+        bigramHashTable = new long[PRIME_BIGRAM_LENGTH];
+        frequencyTable = new int[PRIME_BIGRAM_LENGTH];
+        for (int i = 0; i < PRIME_BIGRAM_LENGTH; i++) {
+          // it is possible for a value to hash to 0, but the probability is extremely low
+          bigramHashTable[i] = 0;
+          frequencyTable[i] = 0;
+        }
+        loadFromFile(bigramDictPath);
+      } catch (IOException e) {
+        throw new RuntimeException(e.getMessage());
+      }
+      saveToObj(serialObj);
+    }
+  }
+
+  /**
+   * Load the datafile into this BigramDictionary
+   * 
+   * @param dctFilePath path to the Bigramdictionary (bigramdict.dct)
+   * @throws IOException If there is a low-level I/O error
+   */
+  public void loadFromFile(String dctFilePath) throws IOException {
+
+    int i, cnt, length, total = 0;
+    // The file only counted 6763 Chinese characters plus 5 reserved slots 3756~3760.  
+    // The 3756th is used (as a header) to store information.
+    int[] buffer = new int[3];
+    byte[] intBuffer = new byte[4];
+    String tmpword;
+    DataInputStream dctFile = new DataInputStream(Files.newInputStream(Paths.get(dctFilePath)));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *dctFile*. The resource is referenced in statements at the following lines: 152, 160, 169, 172, 175, 182. The resource closure statement is at line: 201. There are other execution paths that do not contain closure statements, e.g., when exception is thrown by DataInputStream.read. Either a) close *dctFile* in a try-finally block or b) close the resource by declaring *dctFile* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).